Close open channels can be cleaned up on windows without any warnings.

MishaDemianenko opened the pull request:
Close tests resources. 
Update StoreLocker to close channel in case of failure.
Properly close lock channel in the case when the store was already locked in the batch inserter.
Try to stop lifecycle component that fails its start to clear all allocated resources. 
Close segments in tests.

MishaDemianenko committed:
Close all test allocated resources so files can be cleaned up on windows without any warnings.

MishaDemianenko committed:
Update store lock checker optimistic assumption and log file test

MishaDemianenko committed:
Stop lifecycle instance that fail to start.
Move throwing of exception outside fo catch block in store lock checker.

MishaDemianenko committed:
Close segments

MishaDemianenko committed:
Close channel after test

MishaDemianenko committed:
Properly shutdown components.
Properly close transaction cursor.

MishaDemianenko committed:
Fix checkstyle

MishaDemianenko commented:
@chrisvest updated

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java:
```diff
@@ -1739,51 +1739,52 @@ public void pagesAddedWithNextWithPageIdMustBeAccessibleWithNoGrowSpecified() th
-        PagedFile pagedFile = pageCache.map( file( "a" ), 20 );
-
-        try ( PageCursor cursor = pagedFile.io( 0, PF_SHARED_WRITE_LOCK ) )
+        try ( PagedFile pagedFile = pageCache.map( file( "a" ), 20 ) )
```
These two try-with-resources could be combined.

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java:
```diff
@@ -2067,81 +2068,87 @@ public void evictionMustFlushPagesToTheRightFiles() throws IOException
-            boolean cursorReady2;
-
-            try ( PageCursor cursor = pagedFile1.io( pageId1, PF_SHARED_WRITE_LOCK ) )
+            try ( PagedFile pagedFile2 = pageCache.map( file2, filePageSize2 ) )
```
These two can also be combined.

tinwelint commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/StoreLockChecker.java:
```diff
@@ -58,10 +59,24 @@ static Closeable check( Path databaseDirectory ) throws CannotWriteException
+                {
+                    try
+                    {
+                        storeLocker.close();
```
We've seen this before from previous attempts to "clean up" around this area, at least three times. We can't seem to do this because then the original lock gets released.

MishaDemianenko replied:
we even recently updated StoreLocker tests suite and they claim that it's not the case but if you can't take a look on them and find where it can be the case that would be great, otherwise those test claims we still have a lock

chrisvest replied:
Maybe it'd be a good idea to run a custom Windows build.

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -96,32 +97,25 @@ public void mustEvictCleanPageWithoutFlushing() throws Exception
-        assertThat( clockArm, is( 1 ) );
-        assertNotNull( tracer.observe( Evict.class ) );
-    }
+            try ( PagedFile pagedFile = pageCache.map( file( "a" ), 8 ) )
```
As with the `PageCacheTest`, there are many places here where try-with-resources clauses can be combined.

chrisvest commented on community/common/src/main/java/org/neo4j/kernel/lifecycle/LifeSupport.java:
```diff
@@ -437,6 +437,14 @@ public void start()
+                    catch ( Throwable se )
+                    {
+                        e.addSuppressed( se );
+                    }
```
Should we do the same for `init`?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/RemoteStore.java:
```diff
@@ -103,30 +103,27 @@ private long getPullIndex( File storeDir ) throws IOException
-            TransactionCursor cursor;
-            try
+            try ( TransactionCursor cursor = txStore.getTransactions( lastCleanTxId ) )
             {
```
I think these two could be combined as well.

chrisvest reviewed (commented):
I have a few comments, and TestBackup is failing for some reason. It mostly looks good, though.

MishaDemianenko committed:
introduce GlobalStoreLocker to enforce single channel per locked file
across all the locker instances.

Use relaxed locker for ephemeral database to prevent cases when one non closed db will fail all the consequent tests for that VM.

Close databases across various tests that where expose by global lock holding investigation.

MishaDemianenko committed:
Merge try blocks

MishaDemianenko committed:
Try to shutdown lifecycle instance component if init failed
to shutdown what ever init started/created.

MishaDemianenko commented:
@chrisvest updated, custom windows was there before rebase's it was an official way to confirm that warnings actually were gone; started another one.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/internal/locker/GlobalStoreLocker.java:
```diff
@@ -0,0 +1,84 @@
+    {
+        lockedFiles.remove( storeLockFile );
+        super.releaseLock();
+    }
```
Did you consider the design with a concurrent map from file to a reference count? I think it ought to be possible to do that fairly easily with the atomic update functions of the concurrent hash map.

chrisvest replied:
This is probably fine in practice, though.

MishaDemianenko replied:
that was the first I did actually, but then switch to a simple set of opened lock files since I think it's a bit simpler and still do the thing

chrisvest replied:
I agree after taking a second look.

chrisvest commented on community/kernel/src/test/java/org/neo4j/graphdb/LabelsAcceptanceTest.java:
```diff
@@ -86,7 +87,9 @@
-    public ImpermanentDatabaseRule dbRule = new ImpermanentDatabaseRule();
+    public final ImpermanentDatabaseRule dbRule = new ImpermanentDatabaseRule();
+    @Rule
+    public final TestDirectory testDirectory = TestDirectory.testDirectory();
```
RuleChain?

MishaDemianenko replied:
i usually follow rule of three to switching to rule chain :)

chrisvest replied:
I use it when I think the order might matter.

MishaDemianenko replied:
that's cause and not even discussable since chain rule should solve exactly that, but it also nice to combine those rules into one that I was referring. 
I do not think that order of them are matter in this test, right?

chrisvest reviewed (approved):
Review done. Looks good. Just a few questions.

MishaDemianenko committed:
Shutdown databases in tests
